File: android/src/main/kotlin/dev/coolswood/openpanel/DeviceInfo.kt
package dev.coolswood.openpanel

import android.content.Context
import android.content.res.Resources
import android.os.Build
import java.util.Locale

/** Version of the Flutter SDK reported in headers; keep in sync with pubspec. */
internal const val SDK_VERSION = "0.1.0"

/**
 * Device and app metadata attached to every event. Collected once, lazily —
 * safe to read from any thread after the first access.
 */
internal class DeviceInfo(private val context: Context) {
    val properties: Map<String, Any?> by lazy { collect() }

    val userAgent: String by lazy {
        val props = properties
        val os = "${props["os_name"]} ${props["os_version"]}"
        val model = props["device_model"] ?: ""
        "OpenPanelFlutter/$SDK_VERSION ($os; $model)"
    }

    private fun collect(): Map<String, Any?> = buildMap {
        put("os_name", "Android")
        put("os_version", Build.VERSION.RELEASE)
        put("device_manufacturer", Build.MANUFACTURER)
        put("device_model", Build.MODEL)
        put("locale", Locale.getDefault().toLanguageTag())
        put("sdk_name", "openpanel_flutter")
        put("sdk_version", SDK_VERSION)
        val metrics = Resources.getSystem().displayMetrics
        put("screen_width", metrics.widthPixels)
        put("screen_height", metrics.heightPixels)
        try {
            val info = context.packageManager.getPackageInfo(context.packageName, 0)
            put("app_version", info.versionName)
            put(
                "app_build",
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    info.longVersionCode
                } else {
                    @Suppress("DEPRECATION")
                    info.versionCode.toLong()
                },
            )
        } catch (_: Exception) {
            // Package info is unavailable (rare test environments) — skip.
        }
    }
}

File: ios/openpanel/Sources/openpanel/DeviceInfo.swift
import Foundation
import UIKit

/// Device and app metadata attached to every event.
///
/// `properties` and `userAgent` are collected lazily on first access — the
/// first access happens on the main thread during `initialize` (UIScreen
/// and UIApplication are main-thread-only), afterwards the cached values
/// are safe to read from any thread.
final class DeviceInfo {
  /// Version of the Flutter SDK reported in headers; keep in sync with pubspec.
  static let sdkVersion = "0.1.0"

  private(set) lazy var properties: [String: Any] = collect()
  private(set) lazy var userAgent: String = {
    let props = properties
    let os = "\(props["os_name"] ?? "") \(props["os_version"] ?? "")"
    let model = props["device_model"] ?? ""
    return "OpenPanelFlutter/\(DeviceInfo.sdkVersion) (\(os); \(model))"
  }()

  /// Whether the application is currently visible; main-thread only.
  var isAppActive: Bool {
    let state = UIApplication.shared.applicationState
    return state == .active || state == .inactive
  }

  private func collect() -> [String: Any] {
    let device = UIDevice.current
    let bundle = Bundle.main
    var props: [String: Any] = [
      "os_name": device.systemName,
      "os_version": device.systemVersion,
      "device_model": machineIdentifier(),
      "locale": Locale.current.identifier,
      "sdk_name": "openpanel_flutter",
      "sdk_version": DeviceInfo.sdkVersion,
    ]
    let screen = UIApplication.shared.connectedScenes
      .compactMap { ($0 as? UIWindowScene)?.screen }
      .first
    if let screen {
      props["screen_width"] = Int(screen.bounds.width * screen.scale)
      props["screen_height"] = Int(screen.bounds.height * screen.scale)
    }
    if let version = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String {
      props["app_version"] = version
    }
    if let build = bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String {
      props["app_build"] = build
    }
    return props
  }

  /// Hardware model identifier like "iPhone17,1" — more precise than
  /// `UIDevice.model` which only returns "iPhone"/"iPad".
  private func machineIdentifier() -> String {
    var systemInfo = utsname()
    uname(&systemInfo)
    let mirror = Mirror(reflecting: systemInfo.machine)
    let bytes = mirror.children.compactMap { child -> UInt8? in
      guard let value = child.value as? Int8 else { return nil }
      return value != 0 ? UInt8(value) : nil
    }
    return String(bytes: bytes, encoding: .utf8) ?? UIDevice.current.model
  }
}
